batcher/batcher.go

```diff
@@ -0,0 +1,266 @@
+	cfg                *Config
+	wallet             lndclient.WalletKitClient
+	sendOutputChan     chan *sendOutputRequest
+	numPendingOutsChan chan *numPendingOutsRequest
+}
+
+// New initiates a new Batcher with the given config.
+func New(cfg *Config, walletKit lndclient.WalletKitClient) *Batcher {
+	return &Batcher{
+		cfg:                cfg,
+		wallet:             walletKit,
+		sendOutputChan:     make(chan *sendOutputRequest),
+		numPendingOutsChan: make(chan *numPendingOutsRequest),
+	}
+}
+
+// Run is the main Batcher routine, responsible for handling incoming send
+// requests, collecting them into batches and broadcast. It will run until the
+// passed context is done.
+func (s *Batcher) Run(ctx context.Context) {
+	// sendRequests contain the current set of pending sends we want to batch.
+	var sendRequests []*sendOutputRequest
+
+	// txPublishTicker represents the time between each time we'll check
+	// whether it is time to batch and send the current set of send
+	// requests.
+	s.cfg.TxPublishTicker.Resume()
+	defer s.cfg.TxPublishTicker.Stop()
+
+	for {
+
+		select {
+		// A new send is being requested. Add it to rou set of requests
+		// and start the ticker.
+		case req := <-s.sendOutputChan:
+			// Add the new request to our slice, and sort it.
+			sendRequests = append(sendRequests, req)
+			sort.Slice(sendRequests, func(i, j int) bool {
+				return sendRequests[i].cutoff.Before(
+					sendRequests[j].cutoff,
+				)
+
+			})
+
+			// TODO(halseth): add to existing mempool tx by RBF.
+			s.cfg.TxPublishTicker.Resume()
+
+		case req := <-s.numPendingOutsChan:
+			req.respChan <- len(sendRequests)
+
+		// The ticker ticks, so we'll check whether it is time to
+		// publish a new batch.
+		case now := <-s.cfg.TxPublishTicker.Ticks():
+			// Before continuing, we first check if any of the
+			// requests have had their context expired, if so we
+			// can remove them immediately.
+			i := 0
+			for _, req := range sendRequests {
+				select {
+				// Context done, report an error and ignore it.
+				case <-req.ctx.Done():
+					req.errChan <- fmt.Errorf("context " +
+						"expired before publication")
+					continue
+				default:
+				}
+
+				sendRequests[i] = req
+				i++
+			}
+			sendRequests = sendRequests[:i]
+
+			if len(sendRequests) == 0 {
+				continue
+			}
+
+			// Check if any of the send requests have reached the
+			// time cutoff. Since they are sorted, it is enough to
+			// check the first one.
+			if now.Before(sendRequests[0].cutoff) {
+				continue
+			}
+
+			// At least one of the send requests have reached its
+			// cutoff time. We'll batch everything we have up to
+			// this point, and publish a tx.
+			var txOuts []*wire.TxOut
+			for i := range sendRequests {
+				// If we reach the maximum batch size, we'll
+				// break here. The rest will be sent in the
+				// next batch.
+				if uint32(i) == s.cfg.MaxBatchSize {
+					break
+				}
+
+				txOuts = append(txOuts, sendRequests[i].txOut)
```

Comment:
Isn't this nicer with a slice operation? First decide the number of items (min(maxbatchsize, n)) and then slice in one operation?